File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LocationTracker.kt
package expo.modules.vescapecore.location

import expo.modules.vescapecore.protocol.LocationSnapshot

import android.content.Context
import android.location.Location
import expo.modules.vescapecore.navigation.NavigationController
import expo.modules.vescapecore.recording.RecordingCoordinator
import expo.modules.vescapecore.telemetry.AppDataRepository
import expo.modules.vescapecore.telemetry.TelemetryPipeline
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

// @parity /modules/vescape-core/ios/location/LocationTracker.swift
internal class LocationTracker(
    private val applicationContext: Context,
    private val appDataScope: CoroutineScope,
    private val emitEvent: (String, Map<String, Any?>) -> Unit,
    private val recordingCoordinator: RecordingCoordinator,
    private val telemetryPipeline: TelemetryPipeline,
) {
    private val recentLocations = ArrayDeque<Map<String, Any?>>()
    private val courseDeriver = GpsCourseDeriver()
    var latestLocation: LocationSnapshot? = null
        private set
    var latestPreciseLocation: LocationSnapshot? = null
        private set
    private var lastGpsPersistedAt = 0L

    /**
     * Where the rider is, for callers that need a position rather than a *good* position —
     * Navigation being the one that matters. Freshness beats accuracy here: a weak indoor fix from
     * a second ago is the right place to start a path from, while the last precise fix can be
     * yesterday's and kilometres away. Precise only stands in when nothing newer exists at all.
     *
     * @parity /modules/vescape-core/ios/location/LocationTracker.swift `riderPosition`
     */
    val riderPosition: LocationSnapshot?
        get() = latestLocation ?: latestPreciseLocation

    /**
     * Ingest one platform fix and return the [LocationSnapshot] it produced, so a caller acts on the
     * fix it handed in rather than re-reading mutable tracker state.
     *
     * @parity /modules/vescape-core/ios/location/LocationTracker.swift `onLocationUpdated`
     */
    fun onLocationUpdated(location: Location): LocationSnapshot {
        val accuracyM = if (location.hasAccuracy()) location.accuracy.toDouble() else null
        val speedMps = if (location.hasSpeed()) location.speed.toDouble() else null
        val bearingDeg = if (location.hasBearing()) location.bearing.toDouble() else null
        val precise = isPreciseGpsFix(location.provider, accuracyM)
        // Approximate fixes never feed the course: they are metres of noise apart and would spin a
        // derived bearing, and they are not what the map's GPS heading mode follows either.
        val course = if (precise) {
            courseDeriver.derive(
                latitude = location.latitude,
                longitude = location.longitude,
                speedMps = speedMps,
                bearingDeg = bearingDeg,
                timestamp = location.time,
            )
        } else {
            null
        }
        val snapshot = LocationSnapshot(
            latitude = location.latitude,
            longitude = location.longitude,
            speedMps = speedMps,
            bearingDeg = bearingDeg,
            accuracyM = accuracyM,
            altitudeM = if (location.hasAltitude()) location.altitude else null,
            timestamp = location.time,
            precise = precise,
            courseDeg = course?.bearingDeg,
            courseSourceTimestamp = course?.sourceTimestamp,
        )
        latestLocation = snapshot
        // Every fix moves Route Progress, approximate ones included: the same rule as
        // [riderPosition], where freshness beats accuracy. The bearing comes off the path rather
        // than off the fix, so a noisy position cannot spin it.
        // @parity /modules/vescape-core/ios/location/LocationTracker.swift `onLocationUpdated`
        NavigationController.get(applicationContext)
            .onFix(snapshot.latitude, snapshot.longitude, snapshot.speedMps)
        if (!snapshot.precise) {
            emitEvent("onLocation", snapshot.toMap())
            return snapshot
        }
        latestPreciseLocation = snapshot
        persistLastGpsLocation(snapshot)
        recentLocations.addLast(snapshot.toMap())
        pruneRecentLocations(snapshot.timestamp)
        emitEvent("onLocation", snapshot.toMap())
        recordingCoordinator.recordLocation(snapshot)
        return snapshot
    }

    fun recentLocations(): List<Map<String, Any?>> = recentLocations.toList()

    fun pruneRecentLocations(nowMs: Long) {
        val oldest = nowMs - telemetryPipeline.recentWindowMs()
        while (recentLocations.isNotEmpty()) {
            val timestamp = (recentLocations.first()["timestamp"] as? Number)?.toLong() ?: break
            if (timestamp >= oldest) break
            recentLocations.removeFirst()
        }
    }

    // @parity /modules/vescape-core/ios/location/LastGpsLocationPersistence.swift `onLocationUpdated`
    private fun persistLastGpsLocation(location: LocationSnapshot) {
        val now = System.currentTimeMillis()
        if (now - lastGpsPersistedAt < LAST_GPS_PERSIST_INTERVAL_MS) return
        lastGpsPersistedAt = now
        appDataScope.launch {
            AppDataRepository.get(applicationContext).updateLastGpsLocation(location.latitude, location.longitude)
        }
    }
}

File: modules/vescape-core/ios/location/LocationTracker.swift
import Foundation

/// App-level fix state. Session teardown leaves it intact; only replay teardown clears it.
/// All calls use the GPS callback's serial execution context.
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LocationTracker.kt
internal final class LocationTracker {
  private(set) var latestLocation: TelemetryLocationCapture?
  private(set) var latestPreciseLocation: TelemetryLocationCapture?
  private(set) var recentLocations: [[String: Any?]] = []
  private let courseDeriver = GpsCourseDeriver()
  private let recentWindowMs: () -> Int64
  private let recordLocation: (TelemetryLocationCapture) -> Void
  private let navigationFix: (TelemetryLocationCapture) -> Void
  private let persistLocation: (TelemetryLocationCapture) -> Void

  init(
    recentWindowMs: @escaping () -> Int64,
    recordLocation: @escaping (TelemetryLocationCapture) -> Void,
    navigationFix: @escaping (TelemetryLocationCapture) -> Void,
    persistLocation: @escaping (TelemetryLocationCapture) -> Void
  ) {
    self.recentWindowMs = recentWindowMs
    self.recordLocation = recordLocation
    self.navigationFix = navigationFix
    self.persistLocation = persistLocation
  }

  /// Freshness beats precision for Navigation; a precise fix only fills an absent latest fix.
  /// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LocationTracker.kt `riderPosition`
  var riderPosition: TelemetryLocationCapture? { latestLocation ?? latestPreciseLocation }

  /// Returns the enriched fix for the remaining consumers and the unchanged onLocation payload.
  /// iOS preserves its existing order: recording, latest fix, Navigation, precise state/persistence.
  /// Android currently records after emitting, and offers only precise fixes to recording.
  /// TODO(android parity): align recording consumer ordering in the Ride Track slice #448.
  /// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/location/LocationTracker.kt `onLocationUpdated`
  func onLocationUpdated(_ incoming: TelemetryLocationCapture) -> TelemetryLocationCapture {
    var location = incoming
    // Approximate fixes must not feed the course deriver.
    if location.precise {
      let course = courseDeriver.derive(
        latitude: location.latitude,
        longitude: location.longitude,
        speedMps: location.speedMps,
        bearingDeg: location.bearingDeg,
        timestamp: location.timestamp
      )
      location.courseDeg = course?.bearingDeg
      location.courseSourceTimestamp = course?.sourceTimestamp
    }
    recordLocation(location)
    latestLocation = location
    navigationFix(location)
    if location.precise {
      latestPreciseLocation = location
      persistLocation(location)
      recentLocations.append(location.map)
      pruneRecentLocations(now: location.timestamp)
    }
    return location
  }

  func pruneRecentLocations(now: Int64) {
    let oldest = now - recentWindowMs()
    recentLocations.removeAll { row in
      guard let timestamp = (row["timestamp"] ?? nil) as? NSNumber else { return false }
      return timestamp.int64Value < oldest
    }
  }

  /// Clear replay-derived fixes and course without resetting the app-lifetime persistence throttle.
  /// Android currently retains replay fixes; its session teardown tracks that difference.
  func clearReplayLocations() {
    latestLocation = nil
    latestPreciseLocation = nil
    courseDeriver.reset()
    recentLocations.removeAll(keepingCapacity: true)
  }
}
